### ISSUE TITLE
v0.8.30: potential regression in LID

### ISSUE THREAD

**Issue by logiclove** [CrispStrobe/CrispASR#409]
I randomly get weird language like *`"be"`* bengali or *`"ru"`* russian when the text is clearly 100% french audible with high volume and transcribed correctly.

Hardware:
```
OS: Ubuntu 26.04 LTS
Hardware: ASRock  B850M Pro-A WiFi 
Firmware: 4.43  Wed 2026-06-24  2month 5d 
Kernel:7.0.0  Threads:419  Mem:4.7Gi / 59Gi
amd-smi
+------------------------------------------------------------------------------+
| AMD-SMI            26.5.0+2b22ab01                                           |
| amdgpu Version:    6.19.14.31400000                                          |
| ROCm Version:      7.14.0                                                    |
| VBIOS Version:     00131149                                                  |
| Platform:          Linux Baremetal                                           |
|-------------------------------------+----------------------------------------|
| BDF                        GPU-Name | Mem-Uti   Temp   UEC       Power-Usage |
| GPU  HIP-ID  OAM-ID  Partition-Mode | GFX-Uti    Fan               Mem-Usage |
|=====================================+========================================|
| 0000:03:00.0  AMD Radeon RX 9070 XT | 0 %      33 °C   0             7/317 W |
|   0       0     N/A             N/A | 0 %      0.0 %           5625/16304 MB |
|-------------------------------------+----------------------------------------|
| 0000:11:00.0    AMD Radeon Graphics | N/A        N/A   0                 N/A |
|   1       1     N/A             N/A | N/A        N/A             13/30430 MB |
+-------------------------------------+----------------------------------------+
+------------------------------------------------------------------------------+
| Processes:                                                                   |
|  GPU      PID  Process Name       GTT_MEM  VRAM_MEM  MEM_USAGE  CU %  SDMA   |
|==============================================================================|
|  No running processes found                                                  |
+------------------------------------------------------------------------------+
```

docker-compose.yml
```
...
    environment:
      CAPS_TTS: 1
      CRISPASR_MODEL: ${CRISPASR_MODEL:-/models/cohere-transcribe.gguf}
      CRISPASR_BACKEND: ${CRISPASR_BACKEND:-cohere}
      CRISPASR_LANGUAGE: ${CRISPASR_LANGUAGE:-auto}
      CRISPASR_PORT: ${CRISPASR_PORT:-8080}
      CRISPASR_AUTO_DOWNLOAD: ${CRISPASR_AUTO_DOWNLOAD:-0}
      CRISPASR_CACHE_DIR: ${CRISPASR_CACHE_DIR:-/cache}
      # -nt (`--no-timestamps`) breaks diarization. Drop it if you want diarization output at all.
      # https://github.com/CrispStrobe/CrispASR/issues/326
      # --diarize-embedder is now the dominant cost at ~52.8 s because it runs one unbatched forward per segment.
      # Batching that is still open. If per-pass-local speaker IDs are acceptable for your use,
      # dropping --diarize-embedder remains much faster today
      CRISPASR_EXTRA_ARGS: ${CRISPASR_EXTRA_ARGS:---vad --frequency-penalty 0.85 --lcs-dedup on --lcs-min-length 4 --split-on-punct -t 12  --lid-backend silero --diarize --diarize-method pyannote --sherpa-segment-model auto --diarize-cluster-threshold 0.5 --diarize-max-speakers 4 }
```

Logs:
```
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: === crispasr container starting ===
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: build-info: (missing /app/build-info.txt — local build?)
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: nvidia-smi not present in PATH (no --gpus, or non-CUDA image)
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: libcuda.so.1 not in ldconfig cache (no GPU passthrough)
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: DRI render nodes found: /dev/dri/renderD128 /dev/dri/renderD129 
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: vulkaninfo:
crispasr      |   	driverVersion      = 26.0.3
crispasr      |   	deviceName         = AMD Radeon RX 9070 XT (RADV GFX1201)
crispasr      |   	driverVersion      = 26.0.3
crispasr      |   	deviceName         = AMD Ryzen 9 9900X 12-Core Processor (RADV RAPHAEL_MENDOCINO)
crispasr      |   	driverVersion      = 26.0.3
crispasr      |   	deviceName         = llvmpipe (LLVM 21.1.8, 256 bits)
crispasr      | [2026-08-29T17:12:31Z] crispasr-docker: launching: crispasr --server --host 0.0.0.0 --port 8080 --cache-dir /cache -m /models/cohere-transcribe.gguf --backend cohere -l auto --vad --frequency-penalty 0.85 --lcs-dedup on --lcs-min-length 4 --split-on-punct -t 12 --lid-backend silero --diarize --diarize-method pyannote --sherpa-segment-model auto --diarize-cluster-threshold 0.5 --diarize-max-speakers 4
crispasr      | crispasr 0.8.30 (git f632edf3, Release) [backends: cpu,vulkan]
crispasr      | cpu-isa: OK — built for AVX512F/AVX2/FMA/F16C/BMI2/AVX; host CPU supports the build's ISA baseline
crispasr      | WARNING: radv is not a conformant Vulkan implementation, testing use only.
crispasr      | ggml_vulkan: Found 2 Vulkan devices:
crispasr      | ggml_vulkan: 0 = AMD Radeon RX 9070 XT (RADV GFX1201) (radv) | uma: 0 | fp16: 1 | bf16: 1 | fp4: 0 | warp size: 64 | shared memory: 65536 | int dot: 1 | matrix cores: KHR_coopmat
crispasr      | ggml_vulkan: 1 = AMD Ryzen 9 9900X 12-Core Processor (RADV RAPHAEL_MENDOCINO) (radv) | uma: 1 | fp16: 1 | bf16: 0 | fp4: 0 | warp size: 32 | shared memory: 65536 | int dot: 1 | matrix cores: none
crispasr      | cohere: backend: Vulkan0
crispasr      | cohere: n_threads param=12 (use COHERE_THREADS=N to override)
crispasr      | cohere: BN folded into conv_dw weights for 48 layers
crispasr      | cohere: kv cache     = 32.0 MiB  (dec_head_dim=128 max_ctx=1024 n_heads=8 n_layers=8)
crispasr      | cohere: model weights = 3944.5 MiB
crispasr      | cohere: compute_meta  = 6913.0 KiB
crispasr      | cohere: 0.50s of digital silence — returning an empty transcript
crispasr      | crispasr-server: warmup completed in 0 ms
crispasr      | crispasr-server: backend 'cohere' loaded, model '/models/cohere-transcribe.gguf'
crispasr      | 
crispasr      | crispasr-server: listening on 0.0.0.0:8080
crispasr      |   POST /inference                  — upload audio (native JSON)
crispasr      |   POST /v1/audio/transcriptions    — OpenAI-compatible API
crispasr      |   POST /load                       — hot-swap model
crispasr      |   GET  /health                     — server status
crispasr      |   GET  /backends                   — list backends
crispasr      |   GET  /v1/models                  — model info
crispasr      |

**Comment by logiclove** [comment]
Please let me know if you need me to add audio samples to reproduce the issue.

**Comment by logiclove** [comment]
Example output **`BE`**:
```
crispasr      | crispasr-server: /v1/audio/transcriptions received 'session.ogg' (22335 bytes)
crispasr      | crispasr[vad]: using cached /cache/ggml-silero-v6.2.0.bin
crispasr      | whisper_vad_segments_from_samples: detecting speech timestamps in 76446 samples
crispasr      | whisper_vad_detect_speech: detecting speech in 76446 samples
crispasr      | whisper_vad_detect_speech: n_chunks: 150
crispasr      | whisper_vad_detect_speech: props size: 150
crispasr      | whisper_vad_detect_speech: chunk_len: 158 < n_window: 512
crispasr      | whisper_vad_detect_speech: vad time = 15.12 ms processing 76446 samples
crispasr      | whisper_vad_segments_from_probs: detecting speech timestamps using 150 probabilities
crispasr      | whisper_vad_segments_from_probs: Final speech segments after filtering: 1
crispasr      | whisper_vad_segments_from_probs: VAD segment 0: start = 1.54, end = 3.45 (duration: 1.91)
crispasr      | crispasr[lid-silero]: using cached /cache/silero-lid-lang95-f32.gguf
crispasr      | crispasr[lid]: detected 'be' (p=-7.455) via silero
crispasr      | crispasr-server: LID -> language = 'be' (silero, p=-7.455)
crispasr      | crispasr-server: LID input = 1.91s speech from 1 VAD slice(s)
crispasr      | crispasr: note: pyannote-seg-3.0.gguf is licensed mit
crispasr      | crispasr[diarize]: using cached /cache/pyannote-seg-3.0.gguf
crispasr      | pyannote_seg: loaded (4 LSTM layers, hidden=128)
crispasr      | crispasr-server: transcribed 4.8s audio in 0.14s (33.1x realtime), format=verbose_json
transcribe    | INFO: httpx [request:#21] HTTP Request: POST http://crispasr:8080/v1/audio/transcriptions "HTTP/1.1 200 OK"
transcribe    | INFO: utils.logging_config [request:#21] {'task': 'transcribe', 'language': 'be', 'duration': 4.778, 'text': "Bonjour, je m'appelle Xavier.", 'segments': [{'id': 0, 'start': 1.62, 'end': 3.45, 'text': "Bonjour, je m'appelle Xavier.", 'temperature': 0.0, 'avg_logprob': -0.009687, 'no_speech_prob': 0.0, 'speaker': 'B', 'words': [{'word': 'Bonjour,', 'start': 1.62, 'end': 2.42}, {'word': 'je', 'start': 2.42, 'end': 2.58}, {'word': "m'appelle", 'start': 2.58, 'end': 2.9}, {'word': 'Xavier.', 'start': 2.9, 'end': 3.45}], 'tokens': [6291, 4990, 13784, 998, 529, 13789, 2910, 1635, 4070, 1046, 785, 13785]}]}
```

**Comment by logiclove** [comment]
Example ouput **`RU`**:
```
crispasr      | crispasr-server: /v1/audio/transcriptions received 'session.ogg' (79011 bytes)
crispasr      | crispasr[vad]: using cached /cache/ggml-silero-v6.2.0.bin
crispasr      | whisper_vad_segments_from_samples: detecting speech timestamps in 263326 samples
crispasr      | whisper_vad_detect_speech: detecting speech in 263326 samples
crispasr      | whisper_vad_detect_speech: n_chunks: 515
crispasr      | whisper_vad_detect_speech: props size: 515
crispasr      | whisper_vad_detect_speech: chunk_len: 158 < n_window: 512
crispasr      | whisper_vad_detect_speech: vad time = 50.03 ms processing 263326 samples
crispasr      | whisper_vad_segments_from_probs: detecting speech timestamps using 515 probabilities
crispasr      | whisper_vad_segments_from_probs: Merged 0 adjacent segments, now have 3 segments
crispasr      | whisper_vad_segments_from_probs: Final speech segments after filtering: 3
crispasr      | whisper_vad_segments_from_probs: VAD segment 0: start = 4.93, end = 7.01 (duration: 2.08)
crispasr      | whisper_vad_segments_from_probs: VAD segment 1: start = 8.03, end = 10.85 (duration: 2.82)
crispasr      | whisper_vad_segments_from_probs: VAD segment 2: start = 11.68, end = 14.27 (duration: 2.59)
crispasr      | crispasr[lid-silero]: using cached /cache/silero-lid-lang95-f32.gguf
crispasr      | crispasr[lid]: detected 'ru' (p=-3.924) via silero
crispasr      | crispasr-server: LID -> language = 'ru' (silero, p=-3.924)
crispasr      | crispasr-server: LID input = 9.34s speech from 1 VAD slice(s)
crispasr      | crispasr: note: pyannote-seg-3.0.gguf is licensed mit
crispasr      | crispasr[diarize]: using cached /cache/pyannote-seg-3.0.gguf
crispasr      | pyannote_seg: loaded (4 LSTM layers, hidden=128)
crispasr      | crispasr-server: transcribed 16.5s audio in 0.35s (46.7x realtime), format=verbose_json
transcribe    | INFO: httpx [request:#28] HTTP Request: POST http://crispasr:8080/v1/audio/transcriptions "HTTP/1.1 200 OK"
transcribe    | INFO: utils.logging_config [request:#28] {'task': 'transcribe', 'language': 'ru', 'duration': 16.458, 'text': "Bonjour, je m'appelle Xavier. Je suis ... [REDACTED]
```

**Comment by CrispStrobe** [comment]
Thanks for the thorough report — the logs with exact LID lines made this quickly diagnosable.

**What's happening.** Not a regression in the LID computation: we verified the C++ port tensor-by-tensor against the upstream silero ONNX (front-end cos 0.99996, all encoder stages ≥ 0.99998), and the upstream model produces the **same wrong answers on the same audio**. The silero-lang95 classifier is simply unreliable on hard inputs — heavily compressed audio (your low-bitrate OGGs; our MP3-coded test clip flips `en`→`yo` while the identical WAV/Opus reads `en` at p≈0.999), very short speech (your first example had 1.9 s), band-limited recordings. The tell is in your logs: `p=-7.455` is the **raw top logit**, and on failures the model's whole logit vector deflates (correct detections sit around −1.1; your failures at −7.5/−3.9) — the softmax then renormalizes noise into a confident-looking wrong pick. Your GPU is not involved: the Vulkan build already runs this classifier on CPU.

**Fixed on main, in the next release:**
1. Silero LID confidence is now reported as a real probability (softmax) instead of a raw logit.
2. An **evidence floor** on the raw top logit (default −2.0, tune/disable with `CRISPASR_SILERO_LID_MIN_LOGIT`): below it the answer is discarded as out-of-domain with a clear stderr note — this floor separates every failure case we measured (≤ −3.35) from every correct one (≥ ~−1.1), where a probability threshold provably cannot (our MP3 failure case still had softmax p=0.58).
3. When silero is inconclusive, LID now **falls back to whisper-tiny** automatically (markedly more robust on artifacted audio) instead of handing the garbage to the pipeline.

**Workarounds on v0.8.30, today:** `--lid-backend whisper` (default, robust on exactly this audio), or `--lid-backend probe` (asks the loaded cohere model itself — it can only answer with a language cohere supports), or pin `-l fr` when the language is known.

If you can attach one failing `session.ogg`, we'll add it to the regression suite to lock this behavior on real-world audio — and knowing the encoder/bitrate producing those files would help (they're small for their durations, which fits the low-bitrate-artifact picture).
